### PR TITLE
New endpoint for LTI1.3 OIDC authentication

### PR DESCRIPTION
**Pull request by marcospri**
Implements the OIDC flow, similar to the oauth callback step:

http://www.imsglobal.org/spec/security/v1p0/#step-2-authentication-request

#### With no registration

- Launch an LTI1.3 assignment: https://hypothesis.instructure.com/courses/319/assignments/2781
- The http://localhost:8001/lti/1.3/oidc returns a 403 and you get the sad H view.

#### Create the registration row

- `make shell`
```python
db.add(
    models.LTIRegistration(
        issuer="https://canvas.instructure.com", 
        client_id="93820000000000200", 
        auth_login_url="https://canvas.instructure.com/api/lti/authorize_redirect",                 
        key_set_url="https://canvas.instructure.com/api/lti/security/jwks", 
        token_url="ttps://canvas.instructure.com/api/lti/authorize"
    )
)
tm.commit()
```

This values come from https://hypothesis.instructure.com/accounts/1/developer_keys# using the admin account `chris.shaw+admin@hypothes.is`

`Hypothesis LTI 1.3 (localhost)`, the "Details" column is the client_id.

The rest of the URLs values are in https://canvas.instructure.com/doc/api/file.lti_dev_key_config.html

Once I test this in a couple more PRs I'll move them to `devdata`.

- Launch the assigment again, https://hypothesis.instructure.com/courses/319/assignments/2781
- You'll also get a sad H view but now the http://localhost:8001/lti/1.3/oidc returned a 302 response
- After that there will be a couple more auth request on the LMS side:

  - https://canvas.instructure.com/api/lti/authorize_redirect
  - https://hypothesis.instructure.com/api/lti/authorize?client_id=93820000000000200

And, finally the lti_launch request:

`http://localhost:8001/lti_launches`

which fails with a 403 as is missing the LTI1.1 auth parameters but you can check the JWT included in `id_token` parameter, which decoded includes all the LTI1.3 launch params:

```
{
  "https://purl.imsglobal.org/spec/lti/claim/message_type": "LtiResourceLinkRequest",
  "https://purl.imsglobal.org/spec/lti/claim/version": "1.3.0",
  "https://purl.imsglobal.org/spec/lti/claim/resource_link": {
    "id": "ff3205ae-1156-4e6f-a777-573549efc83f",
    "description": "",
    "title": "Test assignment",
    "validation_context": null,
    "errors": {
      "errors": {}
    }
  },
  "aud": "93820000000000200",
  "azp": "93820000000000200",
  "https://purl.imsglobal.org/spec/lti/claim/deployment_id": "609:6ed5aff92b5257ab607ec9c629844ea8586fde68",
  "exp": 1647539851,
  "iat": 1647536251,
  "iss": "https://canvas.instructure.com",
  "nonce": "e67d9dde8dfb419c9f683c27dbae572f",
  "sub": "3f25366d-ab34-4307-be21-3ea444559e82",
  "https://purl.imsglobal.org/spec/lti/claim/target_link_uri": "http://localhost:8001/lti_launches?url=https%3A%2F%2Felpais.es",
  "picture": "https://canvas.instructure.com/images/messages/avatar-50.png",
  "email": "eng+canvasteacher@hypothes.is",
  "name": "Hypothesis 101 Teacher",
  "given_name": "Hypothesis 101",
  "family_name": "Teacher",
  "https://purl.imsglobal.org/spec/lti/claim/lis": {
    "person_sourcedid": null,
    "course_offering_sourcedid": null,
    "validation_context": null,
    "errors": {
      "errors": {}
    }
  },
  "https://purl.imsglobal.org/spec/lti/claim/context": {
    "id": "6ed5aff92b5257ab607ec9c629844ea8586fde68",
    "label": "LTI",
    "title": "LTI Test",
    "type": [
      "http://purl.imsglobal.org/vocab/lis/v2/course#CourseOffering"
    ],
    "validation_context": null,
    "errors": {
      "errors": {}
    }
  },
  "https://purl.imsglobal.org/spec/lti/claim/tool_platform": {
    "guid": "VCSy8DAKCFIQYaItpSGyXhOXIlk5A6NWduVVG1u3:canvas-lms",
    "name": "University of Hypothesis",
    "version": "cloud",
    "product_family_code": "canvas",
    "validation_context": null,
    "errors": {
      "errors": {}
    }
  },
  "https://purl.imsglobal.org/spec/lti/claim/launch_presentation": {
    "document_target": "iframe",
    "height": null,
    "width": null,
    "return_url": "https://hypothesis.instructure.com/courses/319/external_content/success/external_tool_redirect",
    "locale": "en",
    "validation_context": null,
    "errors": {
      "errors": {}
    }
  },
  "locale": "en",
  "https://purl.imsglobal.org/spec/lti/claim/roles": [
    "http://purl.imsglobal.org/vocab/lis/v2/institution/person#Instructor",
    "http://purl.imsglobal.org/vocab/lis/v2/institution/person#Student",
    "http://purl.imsglobal.org/vocab/lis/v2/membership#Instructor",
    "http://purl.imsglobal.org/vocab/lis/v2/system/person#User"
  ],
  "https://purl.imsglobal.org/spec/lti/claim/custom": {
    "canvas_course_id": 319,
    "canvas_api_domain": "hypothesis.instructure.com"
  },
  "https://purl.imsglobal.org/spec/lti-ags/claim/endpoint": {
    "scope": [
      "https://purl.imsglobal.org/spec/lti-ags/scope/score",
      "https://purl.imsglobal.org/spec/lti-ags/scope/result.readonly",
      "https://canvas.instructure.com/lti-ags/progress/scope/show",
      "https://purl.imsglobal.org/spec/lti-ags/scope/lineitem.readonly",
      "https://purl.imsglobal.org/spec/lti-ags/scope/lineitem"
    ],
    "lineitems": "https://hypothesis.instructure.com/api/lti/courses/319/line_items",
    "validation_context": null,
    "errors": {
      "errors": {}
    },
    "lineitem": "https://hypothesis.instructure.com/api/lti/courses/319/line_items/105"
  },
  "https://purl.imsglobal.org/spec/lti-nrps/claim/namesroleservice": {
    "context_memberships_url": "https://hypothesis.instructure.com/api/lti/courses/319/names_and_roles",
    "service_versions": [
      "2.0"
    ],
    "validation_context": null,
    "errors": {
      "errors": {}
    }
  },
  "https://purl.imsglobal.org/spec/lti/claim/lti11_legacy_user_id": "969fc294ee26e3f1d0c9714af3351dc4eacfd6df",
  "https://purl.imsglobal.org/spec/lti/claim/lti1p1": {
    "user_id": "969fc294ee26e3f1d0c9714af3351dc4eacfd6df",
    "validation_context": null,
    "errors": {
      "errors": {}
    },
    "resource_link_id": "a7f9ed8cc130c51c0beaf7e1bd7bbbac937e7724"
  },
  "errors": {
    "errors": {}
  },
  "https://www.instructure.com/placement": null
}
```

### CODE REVIEW

File: lms/routes.py [hunk 1]
```diff
@@ -105,3 +105,5 @@ def includeme(config):
     config.add_route("admin.index", "/admin/")
     config.add_route("admin.instances", "/admin/instances/")
     config.add_route("admin.instance", "/admin/instance/{consumer_key}/")
+
+    config.add_route("lti.oidc", "/lti/1.3/oidc")
```
Comment (marcospri) [inline lms/routes.py:109]: This doesn't belong to /api/lti, that seems to be reserved now for LTI APIS (eg grading APIs).

I think starting a new "namespace" here makes sense. Arguably `content_item_selection` and `lti_launches` should also be in /lti but I'll leave them as they are for now.

File: lms/views/lti/oidc.py [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+from lms.services import LTIRegistrationService
+from lms.validation import OIDCRequestSchema
+
+
+@view_config(
+    route_name="lti.oidc",
+    request_method=("POST", "GET"),
+    schema=OIDCRequestSchema,
+)
+def oidc_view(request):
```
Comment (marcospri) [inline lms/views/lti/oidc.py:16]: Not much going on here

- Check we have a registration
- Redirect to `registration.auth_login_url`

It looks a bit longer with the dictionary of values here but I don't think we should hide it away. All the values are very OIDC view specific and won't be used elsewhere.

File: lms/views/lti/oidc.py [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+        # http://www.imsglobal.org/spec/security/v1p0/#step-2-authentication-request
+        "scope": "openid",
+        "response_type": "id_token",
+        "response_mode": "form_post",
+        "prompt": "none",
+        "login_hint": params["login_hint"],
+        "lti_message_hint": params["lti_message_hint"],
+        "client_id": params["client_id"],
+        "state": uuid.uuid4().hex,
+        "nonce": uuid.uuid4().hex,
```
Comment (marcospri) [inline lms/views/lti/oidc.py:46]: These two are basically TODOs.

For nonce, if we want/need to check uniqueness we'll need a new table or some new volatile storage infra.

We might need to use it for LTI1.1 too: https://github.com/hypothesis/lms/blob/4da6be5175660cf361cfada5e5528cd2dd4e3224/lms/services/launch_verifier.py#L115


`state` is optional but we can tackle it in the same spike. Implementation might be simpler.

Comment (jon-betts) [reply lms/views/lti/oidc.py:46]: Yeah, looks like we ignore the nonce elsewhere for the moment.

File: tests/unit/lms/views/lti/oidc_test.py [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+    def test_missing_registration(self, lti_registration, pyramid_request):
+        lti_registration.get.return_value = None
+
+        with pytest.raises(HTTPForbidden):
+            oidc_view(pyramid_request)
+
+        lti_registration.get.assert_called_once_with(
+            issuer=pyramid_request.parsed_params["iss"],
+            client_id=pyramid_request.parsed_params["client_id"],
+        )
```
Comment (jon-betts) [inline tests/unit/lms/views/lti/oidc_test.py:18]: We test the same thing in both tests, but in different ways. I'd probably proved this the once in one place and leave it in the other.

File: tests/unit/lms/views/lti/oidc_test.py [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+                    "client_id": pyramid_request.parsed_params["client_id"],
+                    "state": Any(),
+                    "nonce": Any(),
+                }
+            ),
+        )
+        lti_registration.get.assert_called_once_with(
+            pyramid_request.parsed_params["iss"],
+            pyramid_request.parsed_params["client_id"],
+        )
```
Comment (jon-betts) [inline tests/unit/lms/views/lti/oidc_test.py:50]: In general I think it's nice to make our assertions in the same order as execution as close as possible. So as this happens before the other stuff I'd tend to put this above it.

File: tests/unit/lms/views/lti/oidc_test.py [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+                    "prompt": "none",
+                    "login_hint": pyramid_request.parsed_params["login_hint"],
+                    "lti_message_hint": pyramid_request.parsed_params[
+                        "lti_message_hint"
+                    ],
+                    "client_id": pyramid_request.parsed_params["client_id"],
+                    "state": Any(),
+                    "nonce": Any(),
+                }
+            ),
```
Comment (jon-betts) [inline tests/unit/lms/views/lti/oidc_test.py:45]:
```suggestion
            query={
                "scope": "openid",
                "response_type": "id_token",
                "response_mode": "form_post",
                "prompt": "none",
                "login_hint": pyramid_request.parsed_params["login_hint"],
                "lti_message_hint": pyramid_request.parsed_params["lti_message_hint"],
                "client_id": pyramid_request.parsed_params["client_id"],
                "state": Any(),
                "nonce": Any(),
                "redirect_uri": pyramid_request.parsed_params["target_link_uri"],
            },
```

If we include the `redirect_uri` as well, then we don't need to use the contains matcher.

**Review by jon-betts — APPROVED** [review]
Yep this all seems to work as described, although I had to do some debugging on the instructions.

I'm a little lost as to the actual flow in place here, but I can see the various stages kicking off and succeeding or failing as described.

I think this could do with a little polish to tidy up some very long lines and add some doc-strings. I've made a few suggestions.

**Pull request merged**